✨ Added `mgm-hurry plot_stats` command option

taina0407 opened the pull request:
Added `mgm-hurry plot_stat` command to generate [Quanstats](https://github.com/ranaroussi/quantstats) report for specific backtest and auto generate after using ./mgm-hurry backtest command

Image comparison between native freqtrade plot and quanstats to verify output
![Screen Shot 2021-10-17 at 09 43 51](https://user-images.githubusercontent.com/5900860/137608860-4fe13342-0aef-4571-8ef7-91b960949105.png)

# ToDo:
- My development environment is MacOS, tried to using pipenv but seem is will modify lots of requirement when try to add **quantstats** and **pandas** to pipfile. **So please help me to add these libs to Pipfile and requirement-mgm.txt**
- Currently this command only report based on backtest profit return, in futures maybe we can also calculate average close price to add into it as benchmark to have better overview of the backtest. But this will require more CPU cycles to calculate avg close price and while native freqtrade plot already has that so I decided to not implement that feature yet.

taina-edoctor committed:
Add mgm-hurry plot_stat command option

Rikj000 commented on mgm-hurry:
```diff
@@ -934,6 +941,77 @@ class MGMHurry:
             username=self.monigomani_config.config['username'],
             message=f'🚀 Fresh **{strategy}** Plot Profit Results ⬇️', results_paths=[output_file_path])
 
+    def plot_stat(self, choose_results: bool = True,
```
I think it would be better to rename the command to `plot_stats` because we're plotting multiple + 
Then it will align with the `backtest --plot_stats` sub command *(See Line 828)*

Rikj000 commented on mgm-hurry:
```diff
@@ -934,6 +941,77 @@ class MGMHurry:
+
+        self.logger.info(Color.title(f'👉 Plotting profit graph for BackTest results'))
+        try:
+            import quantstats as qs
```
This import should happen at the very top of the `mgm-hurry` class.

We should also add [the latest version of `quantstats`](https://pypi.org/project/QuantStats/) *(Appears to be `v0.0.43` at this point in time)* in the `Pipfile` & `requirements-mgm.txt`, that way the `installer.sh` & `mgm-hurry install_mgm` + @dependabot will also take care of this dependency.

taina0407 replied:
- Moved import to beginning of the file and run isort to sort imports
- Rename 'plot_stat' func to 'plot_stats'
- Added MGM-Hurry docs for plot_stats
Cheer 🥰

Rikj000 commented on mgm-hurry:
```diff
@@ -934,6 +941,77 @@ class MGMHurry:
+            import quantstats as qs
+            # extend pandas functionality with metrics, etc.
+            qs.extend_pandas()
+        except ImportError:
```
I believe it's also safe to remove these `ImportError`s if we move the imports to the top of the `mgm-hurry` file as following:
```python
import pandas as pd  # noqa
import quantstats as qs
```

After moving these to the top it's also recommended to run `python3 -m isort ./mgm-hurry`,
this will automatically optimize & sort the imports in the class :slightly_smiling_face:

Rikj000 reviewed (changes requested):
Some remarks, but it's mostly looking good to me :smile:
Feel free to also update `Docs-MGM-Hurry` for this new command :slightly_smiling_face:

taina-edoctor committed:
Add pip requirement and audit coding / func name

taina-edoctor committed:
Add mgm-hurry plot_stats to Docs-MGM-Hurry

taina0407 commented:
- Moved import to beginning of the file and run isort to sort imports
- Rename 'plot_stat' func to 'plot_stats'
- Added MGM-Hurry docs for plot_stats
Cheer 🥰

Rikj000 reviewed (dismissed):
Nice work mate! Looks ready for merging to me! :tada:
Just let me do a test on it first :slightly_smiling_face:

Rikj000 commented:
Hmmm there seems to be an issue :thinking: 
I pulled in the latest changes on your `feature/hurry-quantstats` branch,
then installed the latest dependencies manually with: `python3 -m pip install -r ./requirements-mgm.txt` and `quantstats` has been installed correctly.

After that I ran: `python3 ./mgm-hurry plot_stats`, but there I get following error:
```log
                                     _                               
 _ __ ___    __ _  _ __ ___         | |__   _   _  _ __  _ __  _   _ 
| '_ ` _ \  / _` || '_ ` _ \  _____ | '_ \ | | | || '__|| '__|| | | |
| | | | | || (_| || | | | | ||_____|| | | || |_| || |   | |   | |_| |
|_| |_| |_| \__, ||_| |_| |_|       |_| |_| \__,_||_|   |_|    \__, |
            |___/                                              |___/ 

INFO - 💨 💨 💨
       👉 Plotting profit graph for BackTest results
       💨 💨 💨
? Please select the BackTest results you want to use:  backtest-result-2021-10-14_22-37-34.json
Traceback (most recent call last):
  File "/mnt/Windows/Users/Rik/Desktop/Rikj-Home/Projects/MoniGoMani-UserAdditions/freqtrade-mgm/./mgm-hurry", line 1149, in <module>
    fire.Fire(MGMHurry)
  File "/home/rikj000/.local/lib/python3.9/site-packages/fire/core.py", line 141, in Fire
    component_trace = _Fire(component, args, parsed_flag_args, context, name)
  File "/home/rikj000/.local/lib/python3.9/site-packages/fire/core.py", line 466, in _Fire
    component, remaining_args = _CallAndUpdateTrace(
  File "/home/rikj000/.local/lib/python3.9/site-packages/fire/core.py", line 681, in _CallAndUpdateTrace
    component = fn(*varargs, **kwargs)
  File "/mnt/Windows/Users/Rik/Desktop/Rikj-Home/Projects/MoniGoMani-UserAdditions/freqtrade-mgm/./mgm-hurry", line 994, in plot_stats
    daily_returns['balance'] = daily_returns['profit_abs'].cumsum() + starting_balance
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/ops/common.py", line 65, in new_method
    return method(self, other)
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/arraylike.py", line 89, in __add__
    return self._arith_method(other, operator.add)
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/series.py", line 4998, in _arith_method
    result = ops.arithmetic_op(lvalues, rvalues, op)
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/ops/array_ops.py", line 189, in arithmetic_op
    res_values = _na_arithmetic_op(lvalues, rvalues, op)
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/ops/array_ops.py", line 139, in _na_arithmetic_op
    import pandas.core.computation.expressions as expressions
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/computation/expressions.py", line 18, in <module>
    from pandas.core.computation.check import NUMEXPR_INSTALLED
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/core/computation/check.py", line 3, in <module>
    ne = import_optional_dependency("numexpr", raise_on_missing=False, on_version="warn")
  File "/home/rikj000/.local/lib/python3.9/site-packages/pandas/compat/_optional.py", line 106, in import_optional_dependency
    module = importlib.import_module(name)
  File "/usr/lib/python3.9/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/usr/lib/python3.9/site-packages/numexpr/__init__.py", line 46, in <module>
    nthreads = _init_num_threads()
  File "/usr/lib/python3.9/site-packages/numexpr/utils.py", line 145, in _init_num_threads
    log.info('Note: NumExpr detected %d cores but "NUMEXPR_MAX_THREADS" not set, so enforcing safe limit of 8.'%n_cores)
  File "/usr/lib/python3.9/logging/__init__.py", line 1446, in info
    self._log(INFO, msg, args, **kwargs)
  File "/usr/lib/python3.9/logging/__init__.py", line 1587, in _log
    record = self.makeRecord(self.name, level, fn, lno, msg, args,
  File "/mnt/Windows/Users/Rik/Desktop/Rikj-Home/Projects/MoniGoMani-UserAdditions/freqtrade-mgm/monigomani/user_data/mgm_tools/mgm_hurry/MoniGoManiLogger.py", line 55, in makeRecord
    log_record.__dict__['message'] = self.clean_line(log_record.__dict__['message'])
KeyError: 'message'
```

I think the key line of failure would be here:
```log
File "/mnt/Windows/Users/Rik/Desktop/Rikj-Home/Projects/MoniGoMani-UserAdditions/freqtrade-mgm/./mgm-hurry", line 994, in plot_stats
    daily_returns['balance'] = daily_returns['profit_abs'].cumsum() + starting_balance
``` 
Either that or the Logger `clean_line` helper function suddenly got broken due to what the command is outputting,
but I highly doubt that :thinking:
```log
  File "/mnt/Windows/Users/Rik/Desktop/Rikj-Home/Projects/MoniGoMani-UserAdditions/freqtrade-mgm/monigomani/user_data/mgm_tools/mgm_hurry/MoniGoManiLogger.py", line 55, in makeRecord
    log_record.__dict__['message'] = self.clean_line(log_record.__dict__['message'])
KeyError: 'message'
```

Rikj000 committed:
⚡️ Explicitly import & use `numpy`

Rikj000 committed:
⚡️ `Pipfile` include `numpy` & `pandas`

Rikj000 committed:
⚡️ `requirements-mgm.txt` include `numpy` & `pandas`

Rikj000 commented:
@taina0407 can you switch line 55 in the `MoniGoManiLogger.py` on your PR 
- from: `log_record.__dict__['message'] = self.clean_line(log_record.__dict__['message'])`
- to: `log_record.__dict__['msg'] = self.clean_line(log_record.__dict__['msg'])`

I believe that should resolve my errors on my end :slightly_smiling_face:
Also feel free to make me a contributor on your fork, because currently I can't do it myself on this PR due to that :smile:

taina0407 commented:
> @taina0407 can you switch line 55 in the `MoniGoManiLogger.py` on your PR
> 
> * from: `log_record.**dict**['message'] = self.clean_line(log_record.**dict**['message'])`
> * to: `log_record.__dict__['msg'] = self.clean_line(log_record.__dict__['msg'])`
> 
> I believe that should resolve my errors on my end 🙂 Also feel free to make me a contributor on your fork, because currently I can't do it myself on this PR due to that 😄

Added you to contributors!
Btw, I've checked in my branch and in my local, line 55 of MoniGoManiLogger.py already as you want to. Any mistake here ?

https://github.com/taina0407/MoniGoMani/blob/feature/hurry-quantstats/user_data/mgm_tools/mgm_hurry/MoniGoManiLogger.py

taina0407 committed:
Merge branch 'Rikj000:development' into feature/hurry-quantstats

taina0407 committed:
Fix bug MGMLogger throw exception on makeRecord

Replace wrong `log_record.__dict__['message']` with `log_record.__dict__['msg']`

taina0407 commented:
> @taina0407 can you switch line 55 in the `MoniGoManiLogger.py` on your PR
> 
> * from: `log_record.**dict**['message'] = self.clean_line(log_record.**dict**['message'])`
> * to: `log_record.__dict__['msg'] = self.clean_line(log_record.__dict__['msg'])`
> 
> I believe that should resolve my errors on my end 🙂 Also feel free to make me a contributor on your fork, because currently I can't do it myself on this PR due to that 😄

Damn, that github format make me mis understood the request, already updated.
![image](https://user-images.githubusercontent.com/5900860/137636915-2c429b3d-fe0f-486c-8613-efe3b2d1dcb5.png)

Rikj000 committed:
⚡️ Fixed indentation

Rikj000 reviewed (approved):
Working on my end! :partying_face: I think we're ready to merge! :smile:
